{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "straight-theater",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import json\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "developmental-cleaners",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_unique(json_file):\n",
    "    with open(json_file) as input_file:\n",
    "        data= json.load(input_file)\n",
    "\n",
    "    for i in range(len(data['cov'])):\n",
    "        sentence= data['cov'][i]['doc']\n",
    "        st= ' '.join(sentence)\n",
    "        data['cov'][i]['doc']= st\n",
    "\n",
    "    res= list(map(dict, set(tuple(sorted(sub.items())) for sub in data['cov'])))\n",
    "    \n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "comparative-arnold",
   "metadata": {},
   "outputs": [],
   "source": [
    "c1= get_unique('covid1.json')\n",
    "c2= get_unique('covid2.json')\n",
    "c3= get_unique('covid3.json')\n",
    "c4= get_unique('covid4.json')\n",
    "c5= get_unique('covid5.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "amino-point",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "58\n",
      "58\n",
      "58\n",
      "100\n",
      "99\n",
      "373\n"
     ]
    }
   ],
   "source": [
    "all_di= [c1,c2,c3,c4,c5]\n",
    "flat_list = [item for sublist in all_di for item in sublist]\n",
    "\n",
    "covid400= {'covid400': flat_list}\n",
    "#print(all_di[0])\n",
    "print(len(c1))\n",
    "print(len(c2))\n",
    "print(len(c3))\n",
    "print(len(c4))\n",
    "print(len(c5))\n",
    "print(len(flat_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "wicked-asset",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Open json\n",
    "with open('covid400.json', 'w') as fp:\n",
    "    json.dump(covid400, fp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fossil-grave",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'doc': 'coronavirus disease  covid is a new infectious disease caused by the severe acute respiratory syndrome coronavirus  sarscov. the common clinical features of the covid range from mild respiratory symptoms to severe acute respiratory distress syndrome and may be accompanied with a wide spectrum of gastrointestinal cardiovascular or neurological manifestations  . covid first appeared in december  in wuhan china where the first pneumonia cases of unknown origin were reported  . the disease has spread rapidly around the globe and the world health organization who declared a covid pandemic on march th   . by  december  the number of infected subjects has reached over  million with more than . million deaths worldwide. currently there are over  million active cases with about . of patients in critical condition   . overall covid has caused numerous socioeconomic consequences in every aspect of daily life ranging from great economic loses in all sectors challenges in the healthcare system travel restrictions social distancing and lockdowns   all of which have seriously impacted mental health with higher rates of anxiety depression and stress reported among the general population  . to date several vaccines have been developed to prevent further spreading of covid  . furthermore massive efforts have been undertaken to find effective treatments for those who had already contracted the diseaseat present there are more than  clinical studies related to covid as listed on clinicaltrials.gov database where majority were focused on chloroquine or hydroxychloroquine with or without azithromycin lopinavirritonavir remdesivir or dexamethasone  . except for dexamethasone which led to improvement in survival of hospitalized patients in need for supplemental oxygen   none of the aforementioned drugs showed significant efficacy in ameliorating the covid outcome    . some of the other identified drugs such as ivermectin ibrutinib imatinib or ruxolitinib are currently in different phases of clinical trials but so far none of them has been recommended for treatment of covid by the covid treatment guidelines panel  . only two drugs are currently approved for covid  in october  u.s. food and drug administration fda approved remdesivir as the first drug for the treatment of severe covid cases in need for hospitalization   and in november  the same entity issued an emergency use authorization for the drug baricitinib  . however since these drugs seem to have limited efficiency and serious side effects   it is urgent to identify more potent and safe therapeutics that could significantly decrease mortality and relieve the burden of covid on human health and healthcare systems worldwide. in conventional circumstances the process of developing new drugs and testing their safety and efficacy in clinical trials takes several years and the median cost of this process is almost a billion usd per drug  . the fastest and cheapest way to discover therapeutics for covid is repurposing of already approved drugs. multiple experimental and computational drug repurposing strategies have been developed to discover novel indications for well characterized drugs that had already passed extensive clinical studies  . examples of successful drug repurposing include sildenafiloriginally an antihypertensive drug repurposed for erectile dysfunction thalidomidesedative that is also effective against erythema nodosum leprosum and multiple myeloma or aspirinan analgesic that can also be used to decrease the risk of cardiovascular disease and colorectal cancer  . in the past decade a principle of transcriptomic signature reversion has been increasingly employed as a computational drug repurposing strategy especially in cancer research  . this approach identifies drugs with inverted transcriptomic signatures in relation to the signature of a disease. treatment of patients with such drugs could thus potentially reverse the disease transcriptomic signature presumably ameliorating the disease phenotype as a result  . signature matching of transcriptomic data started in  with the connectivity map cmap project  which further evolved into the library of integrated networkbased cellular signatures lincs   . transcriptomic signature reversion approach using cmap has been extensively used in pharmacogenomics studies. examples of experimentally validated drugs that were identified by this approach include antiepileptic drug topiramate that showed potential for treating inflammatory bowel disease   mtor inhibitor rapamycin shown to induce glucocorticoid sensitivity in acute lymphoblastic leukemia   and ikarugamycin and quercetin shown to reduce inflammation in cystic fibrosis  . even though clinical trials are still missing to provide a definite proofofconcept such approach has demonstrated a potential for drug prioritization in pharmacogenomics research  . several in silico drug repurposing studies based on a transcriptome reversal approach have already been published in the context of covid      . however one common issue in such studies is a complete lack or insufficiency of the criteria for inclusion of datasets in the analysis which may then produce misleading results. careful consideration of biological parameters such as for example host cell tropism of the virus is needed to assess the suitability of each dataset. furthermore high variability among datasets and noise in the transcriptomes often cast doubts on the validity of the results  . combining biological knowledge with bioinformatics approaches is much needed to ensure the validity of such studies and to increase chances that selected drugs will be efficient in suppressing disease symptoms. in this study we used the cmap computational drug repurposing approach to identify drug candidates with the potential to revert host transcriptome alterations triggered by the sarscov virus. specifically we searched publicly available transcriptomic datasets deposited via gene expression omnibus geo  of sarscov infected cells with matching noninfected controls identified differentially expressed genes degs and employed lincs  database to select drugs with transcriptomic signatures opposite to those induced by the sarscov infection. of note we used a rational strategy based on unsupervised machine learning approaches to select biologically relevant datasets i.e. those that exhibited relatively high magnitude of response to viral infection. moreover we filtered degs to obtain those that are shared among multiple datasets so that they reflect a general and robust response to infection which should correspond to a physiological state of an infected organism. the obtained final list of  drug candidates was then characterized using bioand chemoinformatic analyses to provide additional insights into the pathogenesis and viralhost interaction mechanisms. in this study we determined transcriptome alterations in cells infected with sarscov in relation to noninfected controls and used these data to reveal drugs that induce the oppositesense changes in the subset of the transcriptome affected by the virus. we propose that these drugs could reverse the host transcriptomic signature induced by sarscov and are thus potential candidates for treatment of covid. overall study design is depicted in figure  . results  . combining biological knowledge with bioinformatics approaches is much needed to ensure the validity of such studies and to increase chances that selected drugs will be efficient in suppressing disease symptoms. in this study we used the cmap computational drug repurposing approach to identify drug candidates with the potential to revert host transcriptome alterations triggered by the sarscov virus. specifically we searched publicly available transcriptomic datasets deposited via gene expression omnibus geo  of sarscov infected cells with matching noninfected controls identified differentially expressed genes degs and employed lincs  database to select drugs with transcriptomic signatures opposite to those induced by the sarscov infection. of note we used a rational strategy based on unsupervised machine learning approaches to select biologically relevant datasets i.e. those that exhibited relatively high magnitude of response to viral infection. moreover we filtered degs to obtain those that are shared among multiple datasets so that they reflect a general and robust response to infection which should correspond to a physiological state of an infected organism. the obtained final list of  drug candidates was then characterized using bioand chemoinformatic analyses to provide additional insights into the pathogenesis and viralhost interaction mechanisms. in this study we determined transcriptome alterations in cells infected with sarscov in relation to noninfected controls and used these data to reveal drugs that induce the oppositesense changes in the subset of the transcriptome affected by the virus. we propose that these drugs could reverse the host transcriptomic signature induced by sarscov and are thus potential candidates for treatment of covid. overall study design is depicted in figure  . at first publicly available transcriptomes obtained from cells infected with sarscov and their noninfected counterparts were identified. considering heterogeneity in design of the identified studies datasets for the analysis were preselected based on the predefined criteria see materials and methods. as a result the seven transcriptomic datasets were figure s . these datasets were obtained from four cell lines normal human bronchial epithelial cell line nhbe human lung adenocarcinoma alveolar epithelial cell line a with and without overexpressed angiotensinconverting enzyme  ace host receptor to which sarscov binds and enters the cell and human lung adenocarcinoma airway epithelial cell line calu. for three cell lines datasets obtained upon infection with two different multiplicity of infection moi were included in the analysis table  . to provide an additional source of data for nhbe we included a dataset obtained from human bronchial organoids generated from nhbe cell line even though this dataset did not fully match the inclusion criteria the samples were collected for rnaseq analysis  days postinfection instead of  h postinfection. we next identified degs in each infected sample relative to corresponding noninfected control table s figure sa . interestingly we observed high variation in the number of degs among different datasets indicating that some samples are more responsive to sarscov than the others. we thus decided to reduce the list of analyzed datasets to only those with relatively high magnitude of response to the virus in terms of transcriptome changes. to that end we performed pca on all eight transcriptome datasets and found that the distance on the score plot between infected and noninfected counterparts was apparently larger for aace and calu regardless of the moi as compared to a hbo and nhbe samples  figure  . this pca analysis indicated that the magnitude of response to the virus was higher for aace and calu relative to a without ace overexpression hbo and nhbe samples. to additionally evaluate the effect of moi the separate pca analyses were conducted for each cell line infected with two mois  figure sa c. in aace and calu cell lines there were more differences between noninfected and infected samples than between samples treated with different moi indicating high sensitivity of these lines to low quantities of the virus. conversely a cells showed negligible changes in the transcriptome upon infection with low moi of . as compared to high moi of  suggesting high threshold for infection. next differential gene expression analyses were performed for a aace and calu cell lines considering pairs of sarscov infected and noninfected samples with different moi as covariate table s figure sb . combining transcriptomes of different mois for each cell line in one analysis would correspond to a general and robust response to sarscov infection which much better represents physiological situation where different cells are exposed to a range of quantities of viral particles. furthermore pooling two different mois in this kind of analysis contributes to reduction of noise in the data. in order to reduce the noise in the dataset obtained from nhbe cells for which there were no available transcriptomes upon infection with different mois we performed an analogous analysis of nhbe cells pooled with hbos deg analysis and pca given that hbos were generated from nhbe cells table s  figure sd . to create final selection of datasets for further analysis which takes into account the magnitude of response to the virus and the effect of different moi we performed hierarchical clustering on the results of the four differential gene expression analyses  figure s . this analysis yielded two distinct clusters one that contained aace and calu infected with sarscov at both mois and another that included a hbo and nhbe samples. hence we opted to continue our study only on aace and calu cell lines whereby the data for both mois were included. the next objective was to identify the most relevant degs that reflect robust change in the host transcriptomic signature upon sarscov infection in more than one cell to additionally evaluate the effect of moi the separate pca analyses were conducted for each cell line infected with two mois  figure sa c. in aace and calu cell lines there were more differences between noninfected and infected samples than between samples treated with different moi indicating high sensitivity of these lines to low quantities of the virus. conversely a cells showed negligible changes in the transcriptome upon infection with low moi of . as compared to high moi of  suggesting high threshold for infection. next differential gene expression analyses were performed for a aace and calu cell lines considering pairs of sarscov infected and noninfected samples with different moi as covariate table s figure sb . combining transcriptomes of different mois for each cell line in one analysis would correspond to a general and robust response to sarscov infection which much better represents physiological situation where different cells are exposed to a range of quantities of viral particles. furthermore pooling two different mois in this kind of analysis contributes to reduction of noise in the data. in order to reduce the noise in the dataset obtained from nhbe cells for which there were no available transcriptomes upon infection with different mois we performed an analogous analysis of nhbe cells pooled with hbos deg analysis and pca given that hbos were generated from nhbe cells table s  figure sd . to create final selection of datasets for further analysis which takes into account the magnitude of response to the virus and the effect of different moi we performed hierarchical clustering on the results of the four differential gene expression analyses  figure s . this analysis yielded two distinct clusters one that contained aace and calu infected with sarscov at both mois and another that included a hbo and nhbe samples. hence we opted to continue our study only on aace and calu cell lines whereby the data for both mois were included. the next objective was to identify the most relevant degs that reflect robust change in the host transcriptomic signature upon sarscov infection in more than one cell type  figure s . to that end we overlapped degs obtained from the two selected differential gene expression analyses aace and calu infected with low and high moi and selected only the degs that were shared among these datasets. furthermore only degs whose expression consistently changed in the same directioneither up or down were considered. this procedure resulted in the list of  degs table s . the cellular pathways affected by sarscov infection were determined by performing gene ontology go enrichment analysis of these  degs. the analysis showed that sarscov infection upregulates various cellular responses such as immune and neuroinflammatory response pathways and downregulates coldinduced thermogenesis and cholesterol biosynthetic pathways  figure   table s . as expected one of the go categories affected by sarscov infection was host defense against viral infection. since host defense against viral infection represents a beneficial mechanism for the cells and presumably should not be reverted by drugs we excluded a total of  overexpressed genes that fell into this go category table s . the resulting  degs were employed in the subsequent steps of our study table s figure s . figure s . to that end we overlapped degs obtained from the two selected differential gene expression analyses aace and calu infected with low and high moi and selected only the degs that were shared among these datasets. furthermore only degs whose expression consistently changed in the same directioneither up or down were considered. this procedure resulted in the list of  degs table s . the cellular pathways affected by sarscov infection were determined by performing gene ontology go enrichment analysis of these  degs. the analysis showed that sarscov infection upregulates various cellular responses such as immune and neuroinflammatory response pathways and downregulates coldinduced thermogenesis and cholesterol biosynthetic pathways  figure   table s . as expected one of the go categories affected by sarscov infection was host defense against viral infection. since host defense against viral infection represents a beneficial mechanism for the cells and presumably should not be reverted by drugs we excluded a total of  overexpressed genes that fell into this go category table s . the resulting  degs were employed in the subsequent steps of our study table s figure s . to identify drugs capable of inducing inverted transcriptomic signature in the host relative to transcriptional changes triggered by sarscov we performed a lincs connectivity map analysis of the final list of degs upon sarscov infection  figure  s . although it included data obtained in as many as  cell lines at least for some compounds lincs does not provide data specifically for the selected aace and to identify drugs capable of inducing inverted transcriptomic signature in the host relative to transcriptional changes triggered by sarscov we performed a lincs connectivity map analysis of the final list of degs upon sarscov infection  figure  s . although it included data obtained in as many as  cell lines at least for some compounds lincs does not provide data specifically for the selected aace and calu cells. furthermore some drugs display highly variable effects in different cell lines making it difficult to predict how they would affect the two lines of interest. to increase probability that the selected drugs would have analogous effects on aace and calu cell lines we filtered out compounds with documented highly variable effects across multiple cell lines. this additional filtering step led to the retention of drugs with robust effects across multiple cell lines. in addition we focused on already approved drugs with the aim of their repurposing see materials and methods. the final list includes  drug candidates with a potential to reverse transcriptomic signature upon sarscov infection  table  . these drugs meet the following criteria  to evaluate whether  drug candidates share some biological and molecular properties we performed clustering of the drugs based on the following parameters pharmacological class and current indication mechanism of action moa molecular structure and known protein targets tables s and s . in terms of their current indication we found that  of the drugs cluster as antiinfective agents  as neuropsychiatric drugs and  as cardiovascular drugs whereas the remaining drugs are pharmacologically heterogeneous  figure sa . the moas of these  drugs were also heterogeneous  figure sb . the clusters with two or maximum three drugs were bacterial ribosomal subunit inhibitors bacterial topoisomerase ii inhibitors alpha demethylase inhibitors histamine receptor antagonists and dopamine receptor antagonists. furthermore no significant similarity of the molecular structures was found among the drugs with the maximum tanimoto coefficient of . obtained for itraconazole and ketoconazole which is still lower than the threshold of . above which drugs are considered significantly similar  figure s . selected drug candidates are also heterogeneous based on their physicochemical properties table s figure s . grouping of the drugs based on their protein targets reveals no obvious preference for any of the four main drug target groups gprotein coupled receptors gpcr ion channels kinases nuclear receptors table s figure sa . indeed majority of the selected drugs target gpcrs   and most drug targets are membranebound   figure sb  as in the case when the unfiltered list of all existing drugs is considered    . in summary this analysis illustrates that the  drugs with a potential to reverse sarscov transcriptomic signature are highly heterogeneous in terms of their properties with main clusters based on their current therapeutic indication  figure s . in order to find out which specific biological pathways affected by sarscov infection can be reversed by the selected  drug candidates we performed drug and target set enrichment analysis dsea and tsea. biological pathways regulated by these drugs table s were overlapped with the pathways affected by the virus identified in the previous step of this study  figure   table s . the overlap contains the following categories metabolic and biosynthetic process immune system process cellular and tissue developmental process cellular architecture and dynamics signaling pathways and response to stimulus with steroid metabolic process being targeted by almost half of the selected drug candidates figure   table s. we conclude that these pathways in particular the steroid metabolic processes might be the key druggable pathways for the reversal of sarscov transcriptomic signature. pathways and response to stimulus with steroid metabolic process being targeted by almost half of the selected drug candidates figure   table s . we conclude that these pathways in particular the steroid metabolic processes might be the key druggable pathways for the reversal of sarscov transcriptomic signature. in this study we used a rational approach to filter currently available transcriptomic datasets and determine relevant degs upon sarscov infection followed by using this information to identify drug candidates with inverse transcriptomic signature as compared to sarscovinduced transcriptome changes. this work revealed  diverse drug candidates that could potentially reverse sarscov signature through targeting a range of biological pathways including immune response metabolic and biosynthetic process cell differentiation and proliferation and signaling pathways. to increase chances that the obtained drug candidates for repurposing will be in this study we used a rational approach to filter currently available transcriptomic datasets and determine relevant degs upon sarscov infection followed by using this information to identify drug candidates with inverse transcriptomic signature as compared to sarscovinduced transcriptome changes. this work revealed  diverse drug candidates that could potentially reverse sarscov signature through targeting a range of biological pathways including immune response metabolic and biosynthetic process cell differentiation and proliferation and signaling pathways. to increase chances that the obtained drug candidates for repurposing will be efficient in vivo we applied multiple measures of caution during development of our bioinformatics pipeline. this was achieved by introducing several filtering steps on each level of the analysis and these include  selection of transcriptomic datasets obtained from biosamples whose transcriptomes are highly responsive to sarscov infection  reduction of noise in the available transcriptomic data  removal of the genes important for cellular defense against virus from the list of target degs and  removal of the drugs with documented variable effects across different cell lines. at the time when this study was performed several datasets obtained from various cells or tissues were available. however not all tissues or cells are equally affected by sarscov and hence not all datasets have equal value as a source of transcriptomic data. indeed the host cell tropism of sarscov depends on the cellular expression of factors that control viral entry and reproduction. for instance to successfully infect the host cells sarscov requires the presence of angiotensinconverting enzyme  ace and transmembrane serine protease  tmprss at the cellular surface  . moreover different cell types vary in their ability to support production of the new virions  . in line with these findings we observed dramatic differences in transcriptome alterations among several cell types exposed to sarscov. using pca and hierarchical clustering on the selected cell types figure   figure s  based on their total transcriptomes and their degs we were able to select the two cell linesa cells expressing ace and calu cells for further analyses. these cell lines were relatively sensitive to the virus i.e. their transcriptomes were more responsive to viral infection as compared to wildtype a cells nhbe cells and hbo. the selected cell types are presumably more vulnerable to the infection and thus represent priority targets for therapeutic intervention. of note our results are in agreement with the previously published data suggesting that a and nhbe have low or variable levels of ace receptor    and are thus not ideal models for studying sarscov infection. data noise is a common issue in the analysis of transcriptomes obtained from different sources or from limited number of samples. correspondingly we have observed high variability in degs among different datasets. to reduce the noise we harmonized the transcriptomic data by including only datasets that met a set of defined criteria and by filtering datasets as described above. finally only consistent degsthose that were shared between the two selected biosamples were used in further analyses. upon viral infection many biological pathways are hijacked by the virus and used for production of new virions. however cellular defense against the virus is also activated. reversal of the expression of genes that belong to the latter pathway would obviously represent a disadvantage for the cell at least in the early stages of infection. therefore we performed a gene ontology enrichment to classify all affected degs into various biological pathways. to ensure that the selected drugs will not affect cellular defense against virus we omitted all genes that fell into this category from the list of degs used for drug selection. to identify drugs that could be repurposed for covid we used the filtered subset of degs as an input for lincs connectivity map analysis. since lincs does not contain data specifically for the selected cell lines and many drugs display cell typespecific effects an additional parameter was introduced in the analysis the robustness of drug effects across multiple cell types. in that we eliminated drugs for which we found evidence that their effects are not conserved across multiple cell lines. this step is important as it increases the likelihood that the selected drugs will have the desired effect in other cell types including our biosamples of interest. the above described pipeline figure   resulted in the list of  drug candidates with potential to reverse sarscov transcriptomic signature. however bioand chemoinformatic analysis of these drugs revealed that they are diverse in terms of their chemical structure physicochemical properties targets and biological pathways that they affect which is in agreement with our observation that degs upon sarscov infection are involved in multiple cellular processes. therefore it is possible that each small cluster of drugs could reverse a different subset of sarscov transcriptome signature. it is tempting to speculate that a combination of two or more individual drugs belonging to different clusters could have more potent transcriptome reversal effects as compared to using a single drug. we identified several biological processes affected by sarscov that can be targeted by drugs and these include metabolic developmental immune and signaling processes  figure  . interestingly steroid metabolic process was at the top of these processes as it was targeted by half of the selected drugs. this result is in agreement with a documented role of cholesterol in the infection of the cells by another coronavirus transmissible gastroenteritis virus   as well as by a porcine nidovirus  . furthermore cholesterol is an important constituent of the cellular membranes and those are essential for almost all aspects of the viral life cycle including the attachment of the virus to the cell surface fusion of the virus with the plasma membrane andor endosomes viral replication in doublemembrane vesicles and budding of the virus from intracellular membrane compartments  . finally steroids have a substantial effect on host immune response    . some of the drugs from our list are already being tested for their effects against covid. these include antiinfective drugs ritonavir azithromycin atovaquone and itraconazole antineoplastic drug imatinib and antidepressant drug fluoxetine  . furthermore azithromycin one of the drug candidates selected in this study was previously identified by a recently published networkbased approach  . also at least three receptors that are targeted by drugs identified in this study were also suggested as putative targets in other networkbased approaches. these include sigma nonopioid receptor  sigmar   targeted by nortriptyline and beta adrenergic receptor adrb and androgen receptor ar  targeted by nortriptyline levobunolol and ketoconazole. our results lend support for further investigation of these drugs or drug targets in experimental approaches to treatment of covid. finally this study suggests novel drug candidates for covid treatment such as memantine ibutilide or trimethadione. in comparison with other studies which employed similar computational approach for drug repurposing based on transcriptome reversal      we observed only a minor or no overlap of the drug candidate lists. shared candidates were adrab antagonists nortriptyline in our study as well as ace inhibitor perindopril and nri agonists econazole and ritonavir in our study that were also identified by elhachem et al.  . this limited agreement between similar studies may stem from using different starting transcriptomic datasets as well as from differences in the criteria applied in the dataset and degs selection procedure. the main strengths of this study are a sound study design and integration of current biological knowledge with rigorous statistics. the pipeline we developed employs a rational and biologically relevant selection of the datasets and differentially expressed genes with the aim to increase reliability of the results. a limitation of this work is that it has been performed in the early phase of covid investigations on a relatively small number of available datasets. moreover the selected datasets were obtained from cancer cell lines which are not an optimal source of transcriptomic data for studying a cancerunrelated disease such as covid. nevertheless our pipeline could be applied in future more comprehensive studies upon publication of transcriptomic datasets obtained from more relevant biosamples such as sarscovinfected primary human cell lines. this approach would also benefit from more profound understanding of the cellular tropism of sarscov and of vulnerability of different primary cell lines to the virus. in such future study our pipeline could be refined to incorporate an additional filtering step of the datasets with positive selection of the vulnerable cell lines and negative selection of the indifferent cell lines whereas drugs could be additionally filtered based on their selective efficiency only in the relevant cell types to avoid bystander toxicity. the pipeline could also be further upgraded to address more complex questions such as temporal dimension of transcriptome changes upon sarscov infection and timing of drug treatments. this would require more knowledge on the dynamic nature of cellular changes postinfection. larger number of the relevant transcriptomic datasets could then be analysed after their initial clustering based on time postinfection and cell type. finally while in this study we focused only on robustly affected genes with fold change higher than  due to the limited number of datasets the developed pipeline could also be further refined by optimizing fold change threshold in a genespecific manner given that fold change in expression does not have equal biological effects for all genes. this will also be possible upon generation of more transcriptomic datasets and of more profound knowledge about genes and pathways that are key to the pathogenesis of covid. publicly available transcriptomic datasets were identified by the extensive search of gene expression omnibus geo repository httpswww.ncbi.nlm.nih.govgds  using keywords sarscov or covid initial screening was performed on  july  with a followup on  october . datasets for the analysis were preselected based on the following criteria  human samples  the complete bioinformatics pipeline was performed in the free software environment for statistical computing r version ..  . differential gene expression analysis was performed with the r package deseq version ..  . raw counts from each of the included transcriptomic datasets were first prefiltered to remove genes with read counts lower than . the remaining raw counts were normalized using deseq variance stabilizing transformation vst. pca analysis was performed on the normalized raw counts. for further downstream analysis only degs with false discovery rate fdr adjusted pvalue < . and fold change  for upregulated genes or <. for downregulated genes were considered. hierarchical clustering of datasets was performed with degs as an input with euclidean distance measure and complete linkage as a clustering method using base r function hclust. r package biomart version ..   with ensembl database was used to convert gene names to entrez id for downstream analysis. functional enrichment analysis was performed with the r package clusterprofiler version ..  . go overrepresentation test was done separately for upand downregulated degs and the results were filtered based on fdr adjusted pvalue less than .. redundant go terms were removed by applying semantic similarity method implemented within the function simplify using the similarity cutoff of .  . transcriptomic signatures induced by sarscov infection were compared with the signatures induced by treatments with various small molecule compounds using the cmap analysis approach. the cmap analysis was conducted using lincs reference database phase  via an r package signaturesearch version ..  . within signaturesearch lincs reference database consisted of differential gene expression analysis of  genes obtained upon treatments of  cell lines with  compounds as perturbagens which corresponded to a total of  signatures  . the results of lincs analysis are lists of perturbagencell line connectivity scores represented by tau tau is a standardized score ranging from − to  where more negativepositive value signifies more exten sive reversalenhancement of transcriptomic signature by a perturbagen in a given cell line  . the obtained list of signatures was further filtered according to the following pipeline  fdr adjusted pvalue of weighted connectivity score was given for each perturbagencell line combination. only significant combinations with fdr adjusted pvalue less than . were selected.  tau connectivity score was given for all significant perturbagencell line combinations. wherever a perturbagen was tested in multiple cell lines the mean tau connectivity score and its coefficient of variation cv described as the standard deviation divided by the mean were calculated. only perturbagens with cv <  i.e. those that showed coherent transcriptomic signature in multiple cell lines were chosen. finally all perturbagens with tau < − were filtered for further analysis. the recommended tau threshold of − was lowered to − to increase the final number of identified drug candidates.  the list of perturbagens was additionally reduced to include only approved drugs which were used for downstream analysis. information about drug approval status was obtained via clue repurposing app httpsclue.iorepurposingapp selection of  drugs in launched phase. information about drugs molecular formula molecular structure as canonical simplified molecular input line entry system smiles chemical class pharmacological class current indication based on the anatomical therapeutic chemical atc classification mechanism of action moa and cellular location was collected from pubchem online database httpspubchem.ncbi.nlm.nih.gov  . physicochemical profiles of the drugs were estimated using admet predictor tm . software simulations plus inc. usa with canonical smiles of compounds as inputs  . parameter relative polar surface area relpsa was calculated using datawarrior software  . ionisation states of the drugs were estimated from acidity and basicity ionization constants calculated by admet predictor tm . software. information on drug target and the type of drugtarget interaction was obtained from online databases drugbank httpsgo.drugbank.com  and drug gene interaction database dgidb httpswww.dgidb.org  . cellular location of drug targets was extracted from drugbank. information about drug target protein families and superfamilies was obtained from uniprotkb httpswww.uniprot.org  and interpro httpswww.ebi.ac.ukinterpro   while information on enzyme class was obtained from integrated relational enzyme database intenz httpswww.ebi.ac.uk intenz  . functional enrichments on the levels of drugs drug set enrichment analysis dsea and targets target set enrichment analysis tsea were performed with signaturesearch using hypergeometric test function and go annotation. results were filtered based on fdr adjusted pvalue less than . and redundant go terms were removed using revigo online tool httprevigo.irb.hr  with similarity cutoff of .. clustering of the drugs was performed in the following steps  for structural similarity only canonical smiles were transposed into circular ecfp extendedconnectivity fingerprint of diameter  fingerprints using r package rcdk version ..  with default options  similarity matrix was calculated from binary or ecfp in case of structural similarity fingerprints with default tanimoto similarity metric using package fingerprint version ..    hierarchical clustering was performed using base r function hclust with distance matrix as input  tanimoto similarity metric and default option of complete linkage as a clustering method. all figures except pipelines and drugtargetpathway network were designed in r version ..  using the following packages ggplot version .. to visualize results of pca analysis and create barplots   dendextend version .. to visualize results of hierarchical clustering as dendrogram   and clusterprofiler version .. for depicting results of go enrichment analysis  . drugtargetpathway network was visualized using open source software for network visualization cytoscape version ..  . supplementary materials the following are available online at httpswww.mdpi.com s figure s  selection of the relevant datasets detailed pipeline figure s  minor portion of degs is shared among multiple datasets figure s  the pca score plots for the three cell lines with two different mois and for a combination of nhbe cells and hbo figure s  hierarchical clustering of various biosamples based on transcriptomic signature changes upon sarscov infection figure s  selection of the relevant degs detailed pipeline figure s  final list of consensus degs upon sarscov infection figure s  selection of the drugs detailed pipeline figure s  distribution of  repurposable drug candidates with a potential to reverse transcriptomic signature upon sarscov infection based on their properties figure s  hierarchical clustering of  drug candidates based on molecular structure figure s  pca biplot demonstrating heterogeneity of  drugs in physicochemical space figure s  distribution of  drug candidates based on drug target properties figure s  hierarchical clustering of  drug candidates based on combined properties table s  list of degs for each dataset separately × table s  list of degs for each group of datasets separately × table s  list of  degs common between aace and calu table s  list of significantly enriched pathways involved in sarscov infection table  s  description of go biological process categories for which degs were excluded table s  final list of  degs common between aace and calu after exclusion of host defense against viral infection genes table s  characterization of  drug candidates with a potential to reverse transcriptomic signature upon sarscov infection table s  target characterization of  drug candidates table s  physicochemical properties of  drug candidates table s  main drug target protein families distribution comparison for all fda approved drugs and  drug candidates table  s  list of significantly enriched pathways regulated by  drug candidates table s', 'title': 'Identification of 37 Heterogeneous Drug Candidates for Treatment of COVID-19 via a Rational Transcriptomics-Based Drug Repurposing Approach'}\n"
     ]
    }
   ],
   "source": [
    "with open('covid400.json') as input_file:\n",
    "    data = json.load(input_file)\n",
    "    \n",
    "print(data['covid400'][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "brown-laugh",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'regex'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-15-8d6f583b76bd>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[1;32mimport\u001b[0m \u001b[0mnltk\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      2\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0mnltk\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtokenize\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mword_tokenize\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[1;32mdef\u001b[0m \u001b[0mnormalize\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtex\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m     \u001b[1;31m#Lowercase\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\envs\\tf-gpu\\lib\\site-packages\\nltk\\__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m    131\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0mnltk\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mgrammar\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[1;33m*\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    132\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0mnltk\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mprobability\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[1;33m*\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 133\u001b[1;33m \u001b[1;32mfrom\u001b[0m \u001b[0mnltk\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtext\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[1;33m*\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    134\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0mnltk\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtree\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[1;33m*\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    135\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0mnltk\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mutil\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[1;33m*\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\envs\\tf-gpu\\lib\\site-packages\\nltk\\text.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     28\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0mnltk\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmetrics\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mf_measure\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mBigramAssocMeasures\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     29\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0mnltk\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcollocations\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mBigramCollocationFinder\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 30\u001b[1;33m \u001b[1;32mfrom\u001b[0m \u001b[0mnltk\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtokenize\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0msent_tokenize\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     31\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     32\u001b[0m ConcordanceLine = namedtuple(\n",
      "\u001b[1;32m~\\Anaconda3\\envs\\tf-gpu\\lib\\site-packages\\nltk\\tokenize\\__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     64\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     65\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0mnltk\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdata\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mload\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 66\u001b[1;33m \u001b[1;32mfrom\u001b[0m \u001b[0mnltk\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtokenize\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcasual\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mTweetTokenizer\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcasual_tokenize\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     67\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0mnltk\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtokenize\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmwe\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mMWETokenizer\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     68\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0mnltk\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtokenize\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdestructive\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mNLTKWordTokenizer\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\envs\\tf-gpu\\lib\\site-packages\\nltk\\tokenize\\casual.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     36\u001b[0m \u001b[1;31m######################################################################\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     37\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 38\u001b[1;33m \u001b[1;32mimport\u001b[0m \u001b[0mregex\u001b[0m  \u001b[1;31m# https://github.com/nltk/nltk/issues/2409\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     39\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mhtml\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     40\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'regex'"
     ]
    }
   ],
   "source": [
    "import nltk\n",
    "from nltk.tokenize import word_tokenize\n",
    "\n",
    "def normalize(tex):\n",
    "    #Lowercase\n",
    "    #tex= tex.lower()\n",
    "    #Remove punctuation\n",
    "    #tex= re.sub(r'[^\\w\\s\\.]','', tex)\n",
    "    #Remove white spaces\n",
    "    tex= tex.strip()\n",
    "    #Tokenize\n",
    "    tokens= word_tokenize(tex)\n",
    "    textf= nltk.Text(tokens)\n",
    "    \n",
    "    return textf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "proprietary-statement",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#Tokens per article\n",
    "tokens= []\n",
    "for doc in data['covid400']:\n",
    "    tokens.append(normalize(doc['doc']))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
